include/onnxruntime/core/framework/data_types.h

```diff
@@ -497,7 +550,81 @@ class SparseTensorType : public SparseTensorTypeBase {
  private:
   SparseTensorType() {
     using namespace data_types_internal;
-    TensorElementTypeSetter<elemT>::SetSparseTensorElementType(mutable_type_proto());
+    TensorElementTypeSetter<elemT>::SetSparseTensorElementType(this->mutable_type_proto());
+  }
+};
+
+/// Common base-class for all optional types.
+class OptionalTypeBase : public DataTypeImpl {
+ public:
+  static MLDataType Type();
+
+  bool IsOptionalType() const override {
+    return true;
+  }
+
+  const OptionalTypeBase* AsOptionalType() const override {
+    return this;
+  }
+
+  bool IsCompatible(const ONNX_NAMESPACE::TypeProto& type_proto) const override;
+
+  size_t Size() const override {
+    // should never reach here.
+    ORT_NOT_IMPLEMENTED(__FUNCTION__, " is not implemented");
+  }
+
+  DeleteFunc GetDeleteFunc() const override {
+    // should never reach here.
+    ORT_NOT_IMPLEMENTED(__FUNCTION__, " is not implemented");
+  }
+
+  const ONNX_NAMESPACE::TypeProto* GetTypeProto() const override;
+
+  virtual MLDataType GetElementType() const {
+    // should never reach here.
+    ORT_NOT_IMPLEMENTED(__FUNCTION__, " is not implemented");
+  }
+
+  OptionalTypeBase(const OptionalTypeBase&) = delete;
+  OptionalTypeBase& operator=(const OptionalTypeBase&) = delete;
+
+ protected:
+  ONNX_NAMESPACE::TypeProto& mutable_type_proto();
+
+  OptionalTypeBase();
+  ~OptionalTypeBase() override;
+
+ private:
+  struct Impl;
+  Impl* impl_;
+};
+
+template <typename T, typename elemT>
```

Comment:
><typename T, typename elemT> [](http://example.com/codeflow?start=9&length=28)

I am not sure this schema would fit the bill. For example, if we choose to support Maps, they require two template parameters.

Reply:
ONNX Optionals do not support anything other than tensors and seq(tensors). AFAIK, Optional(maps) and optional(seq that are not tensors) aren't on the roadmap either. Keeping these in mind, I kept the design simple. We can enhance this class if ONNX does support optionals that are not tensors and seq(tensors).

We can't declare a kernel that accepts tensors(T). Kernels are resolved using the ONNX inferred type, not the ORT runtime type. The ONNX inferred type for an edge that is an optional tensor will something like this: `optional(tensor(float))`. To this end, we need to have an equivalent MLDataType to capture this. Hence, both template parameters are needed.